amqpiface/interfaces.go

```diff
@@ -0,0 +1,650 @@
+type NotifyFlower interface {
+	/*
+		NotifyFlow registers a listener for basic.flow methods sent by the server.
+		When `false` is sent on one of the listener channels, all publishers should
+		pause until a `true` is sent.
+
+		The server may ask the producer to pause or restart the flow of Publishings
+		sent by on a channel. This is a simple flow-control mechanism that a server can
+		use to avoid overflowing its queues or otherwise finding itself receiving more
+		messages than it can process. Note that this method is not intended for window
+		control. It does not affect contents returned by basic.get-ok methods.
+
+		When a new channel is opened, it is active (flow is active). Some
+		applications assume that channels are inactive until started. To emulate
+		this behavior a client MAY open the channel, then pause it.
+
+		Publishers should respond to a flow messages as rapidly as possible and the
+		server may disconnect over producing channels that do not respect these
+		messages.
+
+		basic.flow-ok methods will always be returned to the server regardless of
+		the number of listeners there are.
+
+		To control the flow of deliveries from the server, use the Channel.Flow()
+		method instead.
+
+		Note: RabbitMQ will rather use TCP pushback on the network connection instead
+		of sending basic.flow.  This means that if a single channel is producing too
+		much on the same connection, all channels using that connection will suffer,
+		including acknowledgments from deliveries.  Use different Connections if you
+		desire to interleave consumers and producers in the same process to avoid your
+		basic.ack messages from getting rate limited with your basic.publish messages.
+
+	*/
+	NotifyFlow(c chan bool) chan bool
+}
+
+// NotifyReturner represents ability to register a listener for return methods.
+type NotifyReturner interface {
+	/*
+		NotifyReturn registers a listener for basic.return methods.  These can be sent
+		from the server when a publish is undeliverable either from the mandatory or
+		immediate flags.
+
+		A return struct has a copy of the Publishing along with some error
+		information about why the publishing failed.
+
+	*/
+	NotifyReturn(c chan amqp.Return) chan amqp.Return
+}
+
+// NotifyBlocker represents ability to register a listener for blocked and unblocked connection.
+type NotifyBlocker interface {
+	/*
+		NotifyBlocked registers a listener for RabbitMQ specific TCP flow control
+		method extensions connection.blocked and connection.unblocked.  Flow control is
+		active with a reason when Blocking.Blocked is true.  When a Connection is
+		blocked, all methods will block across all connections until server resources
+		become free again.
+
+		This optional extension is supported by the server when the
+		"connection.blocked" server capability key is true.
+
+	*/
+	NotifyBlocked(receiver chan amqp.Blocking) chan amqp.Blocking
```

Comment:
FTR, a single channel of notifications is used for both `connection.blocked` and `connection.unblocked`. See #75 for details.